### ISSUE TITLE
Probability estimation broken for pandas > 0.19.2

### ISSUE THREAD

**Issue by sebhoerl** [pgmpy/pgmpy#1035]
The problem had already been reported here, but not solved unfortunately: #1005 

With `pgmpy==0.1.6` the ML estimation of a BN works if `pandas==0.19.2` is installed, but not if the latest `pandas==0.23.4` is present.

Example code:
```python3
import pandas as pd
import numpy as np

from pgmpy.models import BayesianModel
from pgmpy.estimators import MaximumLikelihoodEstimator, BayesianEstimator

df = pd.DataFrame(np.array([
    ["A", "U"],
    ["A", "V"],
    ["B", "U"],
    ["B", "U"],
    ["A", "U"],
    ["B", "V"],
    ["A", "V"],
    ["A", "V"]
]), columns = ["Dim1", "Dim2"])

model = BayesianModel([("Dim1", "Dim2")])
model.fit(df, estimator = MaximumLikelihoodEstimator)

for cpd in model.get_cpds():
    print(cpd)
```

Output `pandas==0.19.2`:
![image](https://user-images.githubusercontent.com/647500/47365641-45620100-d6dc-11e8-903a-14a98531e2d8.png)

Output `pandas==0.23.4`:
![image](https://user-images.githubusercontent.com/647500/47365612-34b18b00-d6dc-11e8-9dd3-972dad8dd29a.png)

**Comment by jilongs** [comment]
Same issue here, took me a morning debugging before I got here...

**Comment by scottfleming** [comment]
This is still an important problem. It has to do with a change in the way that the MultIndex function behaves between older and newer pandas versions. I've illustrated that [here](https://github.com/scottfleming/pgmpy-examples/blob/master/pandas_versioning_issue_illustration.ipynb). I would suggest that this be a high priority fix because (1) the issue affects core functionality of the package (for my specific application the package was unusable until I figured this out) and (2) the anaconda installation of pgmpy automatically downloads pandas 0.23 with pgmpy, so this is going to be a persistent problem for all new downloads via anaconda. I might recommend first changing the anaconda configurations so that downloading pgmpy via your channel downgrades to the appropriate pandas version and then fixing the underlying pandas issue when there's more time. I'm afraid I'm not familiar enough with pandas to give a robust fix via PR, but hopefully identifying the exact location of the bug gets you a few steps closer.

**Comment by marcotama** [comment]
I tested a few versions of Pandas, and the earliest problem happens in 0.21.0.

**Comment by marcotama** [comment]
I believe the issue is caused by [this change](http://pandas.pydata.org/pandas-docs/stable/whatsnew.html#multiindex-constructor-with-a-single-level) introduced in Pandas 0.21.1:

> The MultiIndex constructors no longer squeezes a MultiIndex with all length-one levels down to a regular Index. This affects all the MultiIndex constructors. (GH17178)

I nailed down the problem to these lines in `BaseEstimator`: https://github.com/pgmpy/pgmpy/blob/fb90fb5db4884a7d4ce648c79247d928ffa94854/pgmpy/estimators/base.py#L115-L125

When state counts are calculated in line 117 the result of `.size()` is indexed by both the states of the node itself and its parents. Then `.unstack(parents)` is used to pivot the node parents from index to columns. When only one parent exists, `unstack()` creates an Index; when more than one exist, a MultiIndex is created instead. This has not changed.

What changed, is the behaviour of `pandas.MultiIndex.from_product()`, used in line 124. Previously, the function also returned either an Index or a MultiIndex depending on the number of `iterables` passed, which worked nicely with `.reindex()` at line 125. Since Pandas 0.21.0, `pandas.MultiIndex.from_product()` returns a MultiIndex in all cases which, for reasons I did not dive into, does not work with `.reindex()` anymore.

I put together a fix for this and submitted a pull request #1050. I am still learning how to do pull requests, so please be understanding if I made mistakes. Also, I only tested this briefly: please, run additional tests.

Cheers!

**Comment by ankurankan** [comment]
https://github.com/pgmpy/pgmpy/pull/1050#issuecomment-451470998
